keps/sig-autoscaling/4951-configurable-hpa-tolerance/README.md

```diff
@@ -0,0 +1,794 @@
+
+Since appropriate tolerance values are workload-dependent, this KEP proposes to let users add custom tolerance values to `HorizontalPodAutoscaler` resources, overriding the existing default value when present.
+
+This solution integrates seamlessly with the existing HPA API since it already allows users to [fine-tune the autoscaler behavior](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#configurable-scaling-behavior).
+The exact API recommended here has been previously proposed in [kep-853][] (see [here](https://github.com/kubernetes/enhancements/pull/1234#discussion_r333036990)), but it was then proposed to implement it separately.
+
+[#116984]: https://github.com/kubernetes/kubernetes/issues/116984
+[#125987]: https://github.com/kubernetes/kubernetes/issues/125987
+[#62013]: https://github.com/kubernetes/kubernetes/issues/62013
+[#aks-3068]: https://github.com/Azure/AKS/issues/3068
+[#keda-1100]: https://github.com/kedacore/keda-docs/issues/1100
+[Kube Control Manager]: https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/
+[kep-853]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/853-configurable-hpa-scale-velocity/README.md
+
+### Goals
+
+- Allow users to override the default workload autoscaling tolerance for specific HPAs.
+
+### Non-Goals
+
+- Allow to customize the cluster-wise tolerance given by Kube Control Manager `--horizontal-pod-autoscaler-tolerance` parameter.
+
+
+## Proposal
+
+We propose to complement the existing `HorizontalPodAutoscaler` `behavior`
+`scaleUp` and `scaleDown` objects with a new field:
+
+- `scaleUp`:
+  - `tolerance`: the tolerance on the ratio between the *current* and *desired* values of the metric this HorizontalPodAutoscaler operates on, for scaling up. operations. Must be greater or equal to 0.
+- `scaleDown`:
+  - `tolerance`: similar to `scaleUp.tolerance`, for scaling down.
+
+Those fields are optional, and default to the value used today.
+
+### Risks and Mitigations
+
+Risks are limited as this proposal does not introduce a new autoscaling parameters, but merely proposes to override an existing parameter for specific HPAs.
+
+A user chosing an inapropriate value only impacts a unique Deployment. This allows users to experiment with different tolerance values before adopting those more widely.
+
+A user disatisfied with the autoscaling behavior following an update of this parameter can fall back on the existing behavior by removing the new field.
+
+## Design Details
+
+The `HorizontalPodAutoscaler` API is updated to add the new `tolerance` field:
+
+```golang
+type HPAScalingRules struct {
+  // tolerance is the tolerance on the ratio between the current and desired
+  // metric value under which no updates are made to the desired number of
+  // replicas.
```

Comment:
Does it make sense to specify that this should be a value between 0 and 1?

Reply:
Indeed there's nothing special with the value 1.

(Arguably it's not useful for scale down as a _down tolerance_ greater than one will always be fulfilled since the metric ratio can't be negative. That said it doesn't hurt, and allows users to set identical up and down tolerances without hitting corner cases.)